pkg/controller/buildrun/buildrun_controller.go

```diff
@@ -232,42 +244,87 @@ func (r *ReconcileBuildRun) Reconcile(request reconcile.Request) (reconcile.Resu
 	ctxlog.Debug(ctx, "starting reconciling request from a BuildRun or TaskRun event", namespace, request.Namespace, name, request.Name)
 
 	buildRun = &buildv1alpha1.BuildRun{}
-	build = &buildv1alpha1.Build{}
 	lastTaskRun := &v1beta1.TaskRun{}
 
 	// for existing TaskRuns update the BuildRun Status, if there is no TaskRun, then create one
-	if err := r.client.Get(ctx, types.NamespacedName{Name: request.Name, Namespace: request.Namespace}, lastTaskRun); err != nil && apierrors.IsNotFound(err) {
+	if err := r.client.Get(ctx, types.NamespacedName{Name: request.Name, Namespace: request.Namespace}, lastTaskRun); err != nil {
+		if apierrors.IsNotFound(err) {
+			err = r.GetBuildRunObject(ctx, request.Name, request.Namespace, buildRun)
+			if err != nil && !apierrors.IsNotFound(err) {
+				return reconcile.Result{}, err
+			} else if apierrors.IsNotFound(err) {
+				// If the BuildRun and TaskRun are not found, it might mean that we are running a Reconcile after a TaskRun was deleted. If this is the case, we need
+				// to identify from the request the BuildRun name associate to it and update the BuildRun Status.
+				r.VerifyRequestName(ctx, request, buildRun)
+				return reconcile.Result{}, nil
+			}
 
-		err = r.GetBuildRunObject(ctx, request.Name, request.Namespace, buildRun)
-		if err != nil && !apierrors.IsNotFound(err) {
-			return reconcile.Result{}, err
-		} else if apierrors.IsNotFound(err) {
-			// If the BuildRun and TaskRun are not found, it might mean that we are running a Reconcile after a TaskRun was deleted. If this is the case, we need
-			// to identify from the request the BuildRun name associate to it and update the BuildRun Status.
-			r.VerifyRequestName(ctx, request, buildRun)
-			return reconcile.Result{}, nil
-		}
+			// The following block is there to cover the scenario where the build controller is restarted for some reason. In this case it reconciles all buildruns. It then
+			// must not create a second TaskRun for a BuildRun. A good way to test this scenario is by running the controller in a cluster where Tekton is NOT running.
+			// Because then the created TaskRun is not getting processed and our BuildRun will not get the LatestTaskRunRef field set.
+			if buildRun.Status.BuildSpec != nil {
```

Comment:
I was able to reproduce this scenario. I didnt need the tekton piece, but only to remove the `build` operator at some point and then instantiate it again. To be really precise, the use case that I can see where a BuildRun gets two TaskRuns happens when(_laying out the steps_):

1. The build operator is running
2. A Build is applied
3. A BuildRun is applied. This generates a `CREATE` event, follow up by a Reconcile where `TaskRun01` is created.
4. The build operator is removed. **Note**: the removal needs to happen before an `UPDATE` event takes place on `TaskRun01`. In other words, at the time the build operator was removed the last seen event was the one in **3.**
5. The build operator is instantiated again.
6. A new `CREATE` event for the BuildRun takes place, this ends-up on `TaskRun02` being generated.

From the above flow, this will not happen if:

During step 4., the build operator is removed only after a `CREATE` event, follow-up by an `UPDATE` event on the BuildRun. 
Is only after the first `UPDATE` event that the `buildRun.Status.LatestTaskRunRef` gets populated. To illustrate why this matters, this PR contains this https://github.com/redhat-developer/build/pull/331/files#diff-ed341b3440a33952b7b12c1cb1f1f970R93-R94 , which basically means "do not try to reconcile on a BuildRun CREATE event if the BuildRun have a `TaskRunRef`.

@SaschaSchwarze0 the fix for this use case can be addressed by ensuring that https://github.com/redhat-developer/build/blob/master/pkg/controller/buildrun/buildrun_controller.go#L315 happens during a `CREATE` event of the BuildRun. Can you please verify this? then we will dont need this extra logic.

Reply:
Hi @qu1queee you are totally correct with your reproduction step. I only stopped Tekton because it was more reliable to reproduce the issue.

> @SaschaSchwarze0 the fix for this use case can be addressed by ensuring that https://github.com/redhat-developer/build/blob/master/pkg/controller/buildrun/buildrun_controller.go#L315 happens during a CREATE event of the BuildRun. Can you please verify this? then we will dont need this extra logic.

Yeah, I know, I can do another buildrun status update to set the LastTaskRunRef right after I created the taskrun. It now depends on whether we want to reach 100 % that this code always works or only 99.99 %. The "problem" with setting the LastTaskRunRef directly after the taskrun creation is that theoretically this can fail:

1. While the kube api server still let us create the taskrun it for whatever reason refuses us to set the status. Intermittent issue, network hick-up, whatever could happen in a distributed system.
2. The build controller itself or the node on which it runs passes away after the taskrun creation was sent to the API server and before it sends the status update with the LastTaskRunRef

Means that there is still a theoretic chance to have a BuildRun without a LastTaskRunRef but where a TaskRun exists, and the buildrun controller would create another taskrun.

Reply:
This "edge scenario" can also happen during the new code where we `LIST` things, because we still do a CLIENT call. For code simplicity, I will not do that and stick to this approach where we move the TaskRef update after inmediate taskrun creation.

Reply:
Okay, I will change the code. We take the risk then.

Reply:
Done.